pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -96,63 +110,136 @@ func (e *DockerRegistryServiceController) Stop() {
 	}
 
-	return e.defaultDockerURL
+	// returning the same one twice will result in a series of stomp
+	return []string{e.defaultDockerURL}
+}
+
+func (e *DockerRegistryServiceController) getServiceCacheKey() string {
+	// the key made by the informer is: meta.Namespace() + "/" + meta.Name()
+	return e.registryNamespace + "/" + e.registryServiceName
 }
 
 // serviceAdded reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs
 func (e *DockerRegistryServiceController) serviceAdded(obj interface{}) {
-	service := obj.(*api.Service)
-	if service.Name != e.registryServiceName {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
 		return
 	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(service)); err != nil {
-		util.HandleError(err)
+	service := obj.(*api.Service)
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 }
 
 // serviceUpdated reacts to the creation of a docker-registry service by updating all service account dockercfg secrets and
 // changing all interestedURLs, if needed
 func (e *DockerRegistryServiceController) serviceUpdated(oldObj interface{}, newObj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	oldService := oldObj.(*api.Service)
 	newService := newObj.(*api.Service)
 	if newService.Name != e.registryServiceName {
 		return
 	}
-	if e.getServiceLocation(oldService) == e.getServiceLocation(newService) {
-		return
-	}
 
-	if err := e.handleLocationChange(e.getServiceLocation(newService)); err != nil {
-		util.HandleError(err)
+	oldServiceKeys := e.getServiceLocations(oldService)
+	newServiceKeys := e.getServiceLocations(newService)
+	if !reflect.DeepEqual(oldServiceKeys, newServiceKeys) {
+		e.markForChange()
 	}
+
 }
 
 // serviceDeleted reacts to the docker-registry deletion by updating all the generated dockercfg secrets
 func (e *DockerRegistryServiceController) serviceDeleted(obj interface{}) {
+	if !e.servicesSynced() {
+		// skip in this case.  Once the sync is done, we'll bump the change separately
+		return
+	}
+
 	service, ok := obj.(*api.Service)
 	if !ok {
 		return
 	}
-	if service.Name != e.registryServiceName {
-		return
+	if service.Name == e.registryServiceName {
+		e.markForChange()
 	}
 
-	if err := e.handleLocationChange(e.defaultDockerURL); err != nil {
-		util.HandleError(err)
-	}
 }
 
 // handleLocationChange goes through all service account dockercfg secrets and updates them to point at a new docker-registry location
-func (e *DockerRegistryServiceController) handleLocationChange(serviceLocation string) error {
-	e.dockercfgController.SetDockerURL(serviceLocation)
+func (e *DockerRegistryServiceController) handleLocationChange() error {
+
+	service, serviceExists, err := e.serviceCache.GetByKey(e.getServiceCacheKey())
```

Comment:
> truth be told, that doesn't look like much of an improvement, but ok.
>  Add a line note

Actually, not ok.  You want me to iterate over all the namespace hits?  Or you want an indexer that indexes exactly how the cache keys are built?